File: mailer.js
const nodemailer = require('nodemailer');

// With SMTP configured, mail goes out for real. Without it, messages are still
// composed and logged instead of delivered -- so the feature is visible during
// development, and a missing config never takes a request down with it.
const configured = Boolean(process.env.SMTP_HOST);

const transporter = configured
  ? nodemailer.createTransport({
      host: process.env.SMTP_HOST,
      port: Number(process.env.SMTP_PORT) || 587,
      secure: String(process.env.SMTP_SECURE) === 'true',
      auth: process.env.SMTP_USER
        ? { user: process.env.SMTP_USER, pass: process.env.SMTP_PASS }
        : undefined
    })
  : nodemailer.createTransport({ jsonTransport: true });

if (!configured) {
  console.warn('[shop-crm] SMTP not configured — decision emails will be logged, not sent. See .env.example.');
}

function decisionEmail({ shopName, customerName, itemName, status }) {
  const approved = status === 'approved';
  const subject = approved
    ? `Your request for ${itemName} was approved`
    : `Update on your request for ${itemName}`;

  const body = approved
    ? `Hi ${customerName},\n\nGood news — your request for ${itemName} has been approved.\n\nGet in touch with us to arrange the next step.\n\n— ${shopName}`
    : `Hi ${customerName},\n\nThanks for your interest in ${itemName}. We are not able to fulfil this request right now.\n\nDo reach out if you would like to talk about alternatives.\n\n— ${shopName}`;

  return { subject, text: body };
}

// Never rejects: a mail failure must not turn a successful approval into an
// error for the shop owner. Problems are logged and swallowed.
async function sendDecision({ to, shopName, customerName, itemName, status }) {
  if (!to) return { sent: false, reason: 'no email on file' };

  const { subject, text } = decisionEmail({ shopName, customerName, itemName, status });
  const message = {
    from: process.env.SMTP_FROM || `${shopName} <no-reply@localhost>`,
    to,
    subject,
    text
  };

  try {
    const info = await transporter.sendMail(message);
    if (!configured) {
      console.log(`[shop-crm] would email ${to}: ${subject}`);
      return { sent: false, reason: 'smtp not configured', preview: info.message };
    }
    return { sent: true, id: info.messageId };
  } catch (err) {
    console.error(`[shop-crm] email to ${to} failed: ${err.message}`);
    return { sent: false, reason: err.message };
  }
}

module.exports = { sendDecision, decisionEmail, configured };

File: mailer.py
import os
import smtplib
import threading
from email.message import EmailMessage

SMTP_HOST = os.environ.get("SMTP_HOST", "")
SMTP_PORT = int(os.environ.get("SMTP_PORT") or 587)
SMTP_SECURE = os.environ.get("SMTP_SECURE", "").lower() == "true"
SMTP_USER = os.environ.get("SMTP_USER", "")
SMTP_PASS = os.environ.get("SMTP_PASS", "")
SMTP_FROM = os.environ.get("SMTP_FROM", "")

configured = bool(SMTP_HOST)

if not configured:
    print("[shop-crm] SMTP not configured - decision emails will be logged, not sent. See .env.example.", flush=True)


def decision_email(shop_name, customer_name, item_name, status):
    approved = status == "approved"
    if approved:
        subject = f"Your request for {item_name} was approved"
        body = (
            f"Hi {customer_name},\n\n"
            f"Good news - your request for {item_name} has been approved.\n\n"
            f"Get in touch with us to arrange the next step.\n\n"
            f"- {shop_name}"
        )
    else:
        subject = f"Update on your request for {item_name}"
        body = (
            f"Hi {customer_name},\n\n"
            f"Thanks for your interest in {item_name}. We are not able to fulfil "
            f"this request right now.\n\n"
            f"Do reach out if you would like to talk about alternatives.\n\n"
            f"- {shop_name}"
        )
    return subject, body


def _deliver(message, to):
    try:
        if SMTP_SECURE:
            server = smtplib.SMTP_SSL(SMTP_HOST, SMTP_PORT, timeout=20)
        else:
            server = smtplib.SMTP(SMTP_HOST, SMTP_PORT, timeout=20)
            server.starttls()
        with server:
            if SMTP_USER:
                server.login(SMTP_USER, SMTP_PASS)
            server.send_message(message)
    except Exception as err:
        # Never raises to the caller: a mail failure must not turn a successful
        # approval into an error for the shop owner.
        print(f"[shop-crm] email to {to} failed: {err}", flush=True)


def send_decision(to, shop_name, customer_name, item_name, status):
    """Send a decision email. Returns immediately; delivery runs on a thread so
    the owner's approval is not held up by the mail server."""
    if not to:
        return {"sent": False, "reason": "no email on file"}

    subject, body = decision_email(shop_name, customer_name, item_name, status)

    if not configured:
        print(f"[shop-crm] would email {to}: {subject}", flush=True)
        return {"sent": False, "reason": "smtp not configured"}

    message = EmailMessage()
    message["From"] = SMTP_FROM or f"{shop_name} <no-reply@localhost>"
    message["To"] = to
    message["Subject"] = subject
    message.set_content(body)

    threading.Thread(target=_deliver, args=(message, to), daemon=True).start()
    return {"sent": True}
